# ATAC-seq run 161026_M00528_0239_000000000-ANY8K




# ATAC-seq run 161026_M00528_0239_000000000-ANY8K

C1 ATAC-seq run `1772-123-148`.

Example view: <http://fantom.gsc.riken.jp/zenbu/gLyphs/#config=TBD>.


## Processed by hand with BWA.

### Common definitions.

#### Shell variables:

```
BWA=/home/plessy/snap/bwa_0.5.10/usr/bin/bwa
LIBRARY=161026_M00528_0239_000000000-ANY8K
MISEQ=/sequencedata/MiSeq/$LIBRARY/Data/Intensities/BaseCalls
RUN=1772-123-148
```

#### Link farm:

```
mkdir -p ${LIBRARY}
cd $LIBRARY

for SAMPLE in $MISEQ/*_L001_R?_001.fastq.gz
do
  ln -sf $SAMPLE
done

sed 1d /osc-fs_home/scratch/moirai/nano-fluidigm/input/samplename_to_sampleid.txt |
  head -n 96 |
  while read NAME ID
  do
    prename "s/${ID}_S${ID}_L001/${NAME}/" ${ID}_S${ID}_*
  done
prename "s/RunA/$RUN/ ; s/_001//" *z
```

#### Alignment command

```
alignOnGenome() {
  SAMPLE=$1
  GENOME=$2
  $BWA aln -t8 $GENOME -f ${SAMPLE}_R1.sai <(zcat ${SAMPLE}_R1.fastq.gz)
  $BWA aln -t8 $GENOME -f ${SAMPLE}_R2.sai <(zcat ${SAMPLE}_R2.fastq.gz)
  $BWA sampe   $GENOME    ${SAMPLE}_R1.sai ${SAMPLE}_R2.sai \
                           <(zcat ${SAMPLE}_R1.fastq.gz) \
                           <(zcat ${SAMPLE}_R2.fastq.gz) |
    samtools view -uS    |
    samtools sort        |
    tee $SAMPLE.bam      |
    samtools view -u -f2 |
    samtools rmdup - -   |
    samtools sort -n     |
    samtools fixmate - - |
    pairedBamToBed12     |
    awk '{OFS="\t"}{if($6=="+"){$9="0,128,0"};if($6=="-"){$9="128,0,128"};print}' |
    sort -V -k1,1 -k2,2n -k3,3 -k6 > $SAMPLE.bed
  rm ${SAMPLE}_R1.sai ${SAMPLE}_R2.sai
}
```

`pairedBamToBed12` generates many errors because of reads marked _properly
paired_ by BWA on chrM, but where the next mate is maked unmapped (0x8).
This is not corrected by `samtools fixmate`.

### Align on mouse genome

```
GENOME=/analysisdata/genomes/mm9_male.fa

for SAMPLE in $(basename -s _R1.fastq.gz *_R1.fastq.gz)
do
  alignOnGenome $SAMPLE $GENOME
done

prename 's/.bam/_mm9.bam/ ; s/.bed/_mm9.bed/' *[A-H][0-9][0-9].b[ae][md] Undetermined_S0_L001.b*
```

### Align on human genome

```
GENOME=/analysisdata/genomes/hg19_male.fa

for SAMPLE in $(basename -s _R1.fastq.gz *_R1.fastq.gz)
do
  alignOnGenome $SAMPLE $GENOME
done

prename 's/.bam/_hg19.bam/ ; s/.bed/_hg19.bed/' *[A-H][0-9][0-9].b[ae][md] Undetermined_S0_L001.b*
```

### Alignment statistics

#### Proper pairs

```
for bed in *.bed
do
  printf "proper_pairs\t$(basename $bed .bed)\t"
  cat $bed |
  wc -l
done |
  tee ../$LIBRARY.dedupStats.log
```

#### Proper pairs (Q > 20)

```
for bed in *.bed
do
  printf "proper_pairs_Q20\t$(basename $bed .bed)\t"
  awk '$5 > 20' $bed |
  wc -l
done |
  tee ../$LIBRARY.dedupStatsQ20.log
```

#### Redundancy

```
for bam in *.bam
do
  printf "redundancy\t$(basename $bam .bam)\t"
  samtools rmdup $bam  /dev/null 2>&1 |
    grep library |
    cut -f 6 -d ' '
done |
  tee ../$LIBRARY.redundancy.log
```

#### Redundancy (on proper pairs)

``` 
for bam in *.bam
do
  printf "redundancy2\t$(basename $bam .bam)\t"
  samtools view -u -f2 $bam |
  samtools rmdup - /dev/null 2>&1 |
    grep library |
    cut -f 6 -d ' '
done |
  tee ../$LIBRARY.redundancy2.log
```

## Analysis in R

### Library information.


```r
LIBRARY      <- "161026_M00528_0239_000000000-ANY8K"
ZENBU_COLLAB <- "TBD" # C1 CAGE Moore lab collaboration.
RunA         <- "1772-123-148"
```

### Load R packages.


```r
library("magrittr")
library("smallCAGEqc")
library("ggplot2")
library("platetools")
library("viridis")
```

### Create summary table


```r
libs <- data.frame(Well=platetools::num_to_well(1:96), stringsAsFactors = FALSE)
libs$Run <- RunA
libs$Run <- factor(libs$Run)
libs$Row <- sub(".", "", libs$Well) %>% factor
libs$Col <- sub("..$", "", libs$Well) %>% factor
libs$Chamber <- fldgmChamberWell(libs$Well)
rownames(libs) <- paste(RunA, libs$Well, sep="_")
libs$samplename <- rownames(libs)
```

### Load color data


```r
getCuration <- function(colnumber)
  gdata::read.xls( "1772-123-148_cell-count.xlsx"
                 , head = FALSE
                 , stringsAsFactors = FALSE)[[colnumber]]

libs[paste(RunA, fldgmChamberWell( 1:48), sep="_"), "Cell"] <- getCuration(2)
libs[paste(RunA, fldgmChamberWell(49:96), sep="_"), "Cell"] <- getCuration(6)

assignGroup <- function (pattern, group)
  libs[grep(pattern, libs$Cell), "Group"] <<- group

libs$Group <- ""
assignGroup("R",     "mouse")
assignGroup("G",    "human")
assignGroup("GR",    "mix")
assignGroup("empty", "empty")
assignGroup("Dead",  "dead")
libs$Group %<>% factor
```

### Load alignment statistics


```r
loadTriples <- function(pattern='*\\.log') {
  logfiles <- list.files(path='.', pattern=pattern)
  df <- data.frame( variable   = factor()
                  , samplename = factor()
                  , value      = numeric())
  for (logfile in logfiles) {
      df <- rbind(df, read.table(logfile, col.names=colnames(df)))
  }
  df <- reshape::cast(df)
  rownames(df) <- df$samplename
  return(df)
}

alStats <- loadTriples(paste0(LIBRARY, ".*.log"))
alStats$genome <- sub(".*_", "", rownames(alStats))
alStats$samplename %<>% sub(pat="_(mm9|hg19)", rep="")

libs %<>% merge(alStats, by="samplename", all=TRUE)
rm(alStats)
```

### Comparisons


```r
qplot(data = libs, genome, proper_pairs,  geom = "boxplot", color=genome) + ggtitle('Proper pairs') + facet_wrap(~Group, scale="free")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/troubleshoot-1.png)<!-- -->

```r
qplot(data = libs, genome, proper_pairs_Q20,  geom = "boxplot", color=genome) + ggtitle('Proper pairs (Q > 20)') + facet_wrap(~Group, scale="free")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/troubleshoot-2.png)<!-- -->

```r
qplot(data = libs, genome, redundancy,  geom = "boxplot", color=genome) + ggtitle('Redundancy') + facet_wrap(~Group, scale="free")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/troubleshoot-3.png)<!-- -->

```r
qplot(data = libs, genome, redundancy2,  geom = "boxplot", color=genome) + ggtitle('Redundancy (proper pairs only)') + facet_wrap(~Group, scale="free")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/troubleshoot-4.png)<!-- -->

Samples F09 and H02 gave very large number of aligned pairs on the mouse
genome.  For the purpose of better comparing the other samples, the plots
below move them to a "outliers" category.


```r
local({
libs$Group %<>% as.character
libs[grepl("(F09|H02)", libs$samplename), "Group"] <- "outlier"
libs$Group %<>% factor
qplot(data = libs, genome, proper_pairs_Q20,  geom = "boxplot", color=genome) + ggtitle('Proper pairs (Q20)') + facet_wrap(~Group, scale="free")
})
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/troubleshoot-removed-outliers-1.png)<!-- -->


```r
libsM <- libs[libs$genome == "mm9", ][1:96,]
libsH <- libs[libs$genome == "hg19",][1:96,]

raw_map(libsM$Group %>% as.character, well=libsM$Well) +
  ggtitle("Group")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-1.png)<!-- -->

```r
raw_map(libsM$proper_pairs %>% log, well=libsM$Well) +
  ggtitle("Proper pairs (log), mouse") + 
  scale_fill_viridis()
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-2.png)<!-- -->

```r
raw_map(libsH$proper_pairs %>% log, well=libsH$Well) +
  ggtitle("Proper pairs (log), human") + 
  scale_fill_viridis()
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-3.png)<!-- -->

```r
raw_map(libsM$redundancy, well=libsM$Well) +
  ggtitle("Redundancy, mouse") + 
  scale_fill_viridis()
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-4.png)<!-- -->

```r
raw_map(libsH$redundancy, well=libsH$Well) +
  ggtitle("Redundancy, human") + 
  scale_fill_viridis()
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-5.png)<!-- -->

```r
plot(libsM$redundancy, libsM$proper_pairs, log="y", col=libsM$Group %>% as.numeric + 1)
abline(h=1e4, col="red")
abline(v=0.7, col="red")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-6.png)<!-- -->

```r
plot(libsH$redundancy, libsH$proper_pairs, log="y", col=libsH$Group %>% as.numeric + 1)

abline(h=1e4, col="red")
abline(v=0.7, col="red")
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-7.png)<!-- -->

```r
boxplot(data = libs, proper_pairs ~ Group)
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-8.png)<!-- -->

```r
boxplot(data = libs, redundancy ~ Group)
```

![](161026_M00528_0239_000000000-ANY8K_files/figure-html/unnamed-chunk-2-9.png)<!-- -->


Upload libraries on Zenbu
-------------------------


```r
library(zenbuWrappeR)

samplesToUpload <- paste0(libs$samplename, "_", libs$genome)

sampleDescription <- Vectorize(
  function(sample)
    paste(sample, LIBRARY, "ATAC", "knitrUpload", "upload1")
)

sampleUploadTable <-
 data.frame( fullPath    = paste0( getwd()
                                 , "/"
                                 , LIBRARY
                                 , "/"
                                 , samplesToUpload
                                 , ".bed")
           , displayName = samplesToUpload
           , description = sampleDescription(samplesToUpload)
)

write.table(
   sampleUploadTable
 , file = paste0(LIBRARY, ".zUpload.tsv")
 , quote=FALSE
 , sep="\t"
 , row.names = FALSE
 , col.names = FALSE)

ASSEMBLY <- "mm9"
# 
# zenbuUpload( "-assembly", ASSEMBLY
#            , "-filelist", paste0(LIBRARY, ".zUpload.tsv")
#            , "-collab_uuid", ZENBU_COLLAB
#            , "-singletag_exp")

ASSEMBLY <- "hg19"
# 
# zenbuUpload( "-assembly", ASSEMBLY
#            , "-filelist", paste0(LIBRARY, ".zUpload.tsv")
#            , "-collab_uuid", ZENBU_COLLAB
#            , "-singletag_exp")

# Wait for the data to be loaded...

# Need to share again, since the sharing of some samples failed.
# 
# zenbuUpload( "-list"
#            , "-filter"
#            , LIBRARY
#            , "-share"
#            , ZENBU_COLLAB)

tagMetaTable <- function(Key)
 data.frame(
   filter  = sampleDescription(rownames(libs))
 , command = "add"
 , key     = Key
 , value   = libs[[Key]] %>% as.character
 , stringsAsFactors = FALSE
 )

mkMetaTable <- function(Keys) {
 if (! all( Keys %in% colnames(libs)))
   stop("Some keys are not in the libs table")
 Reduce(
     function(DF, Key) rbind(DF, tagMetaTable(Key))
   , Keys
   , data.frame() )
}

metaToTag <- c(
   "Run", "Row", "Chamber"
  , "Group"
  , "genome")

write.table(
   mkMetaTable(metaToTag)
 , file      = paste0(LIBRARY, ".zMeta.tsv")
 , quote     = FALSE
 , sep       = "\t"
 , row.names = FALSE
 , col.names = FALSE)
# 
# zenbuUpload( "-mdfile"
#            , paste0(LIBRARY, ".zMeta.tsv"))
```
